ComplexIndexMap

JoeyT1994 opened the pull request:
This PR adds the ability to use complex arguments within the library by introducing the `ComplexIndexMap` type.

The `ComplexIndexMap` works by containing an additional dictionary, in comparison to the `RealIndexMap`, which keeps track of whether an index is imaginary of real. The `index_to_scalar` function is then overloaded and will return an imaginary vs real number depending on the nature of the index. 

All elementary functions (aside from delta functions -> to be PR'd shortly) now can be constructed as complex versions and this will be automatic based on whether s::IndsNetworkMap wraps a `ComplexIndexMap` or a `RealIndexMap`. Extensive testing is added and example/construct_multidimensional_function.jl has been modified to do complex version of the functions.

Operators (such as differentiation, boundary operators etc) are not tested within this framework and currently assume the underlying `IndexMap` is real (maybe we should add a check for this).

Complex numbers can be accessed by calling `complex_continuous_siteinds(g, args...; kwargs...)` which will build an `IndsNetworkMap` wrapping a `ComplexIndexMap`. One should pass the arguments  `complex_continuous_siteinds(g, real_dimension_vertices::Vector{Vector{V}}, imag_dimension_vertices::Vector{Vector{V}})` to specify where the real indices should be placed and where the imaginary indices should be placed these are anticipated to be ordered by dimension and then digit as usual.
 
There is complete flexibility here over the placement and multiple indices and can be placed on the same vertex.

The library has been updated to the latest version of ITensorNetworks with this PR.

This PR supercedes #32 which I will close.

@ryanlevy @ryboselm @emstoudenmire

ryanlevy commented on src/elementary_functions.jl:
```diff
@@ -133,45 +139,50 @@ by indsnetwork"""
   g_tree = undirected_graph(random_bfs_tree(g, first(vertices(g))))
   s_tree = add_edges(rem_edges(_s, edges(g)), edges(g_tree))
   s_tree = IndsNetworkMap(s_tree, indexmap(s))
+  eltype = indexmaptype(s) == RealIndsNetworkMap ? Float64 : ComplexF64
```
This reminds me, we need to set up the `Adapt` system so we can convert ITN Functions between types

JoeyT1994 replied:
Good point. Yes we should sort that out in a near term PR

ryanlevy reviewed (changes requested):
Thanks, this is great; I like this setup over the previous for sure.

Could you add delta function tests for complex points?
Maybe not this PR if its too complex (hah) but would be nice to have a real -> complex converter (for the TN and index maps). Where you stick an extra complex digit onto each real vertex location and then forming the `ComplexIndexMap` should be straight forward

JoeyT1994 commented on src/IndexMaps/complexindexmap.jl:
```diff
@@ -0,0 +1,122 @@
+      i ->
+        index_dimension(imap)[i] == dim &&
+          index_digit(imap)[i] == digit &&
+          ((ind_type == "Real") == is_real(imap, i)),
```
Yeah that makes sense, will change

JoeyT1994 commented on src/IndexMaps/complexindexmap.jl:
```diff
@@ -0,0 +1,122 @@
+  )
+end
+
+function ComplexIndexMap(
```
Sure thing

JoeyT1994 commented on src/elementary_operators.jl:
```diff
@@ -257,17 +257,21 @@ function operator_proj(fx::ITensorNetworkFunction)
+    sindssim = sim.(sinds)
+    for (i, s) in enumerate(sinds)
+      ssim = sindssim[i]
+      fxgx[v] = fxgx[v] * delta(s, s', ssim)
```
Yeah you're right the `for loop` is unnecessary and this can be done in one line. Although I think the efficiency is pretty much unaffected.

JoeyT1994 commented on src/elementary_functions.jl:
```diff
@@ -32,13 +32,19 @@ function exp_itensornetwork(
+    sinds_dim = filter(i -> dimension(s, i) == dim, sinds)
+    sinds_not_dim = filter(i -> dimension(s, i) != dim, sinds)
+    linds = setdiff(inds(ψ[v]), sinds)
+    ψ[v] = prod([
```
I think it becomes a bit complicated that way because this is constructing an `ITensor` of arbitrary number of physical indices. So it's easiest to write it as an outer product over those indices rather than trying to construct the array first and splat `sinds_dim`

JoeyT1994 commented on src/utils.jl:
```diff
@@ -101,10 +101,57 @@ function base(s::IndsNetwork)
 end
 
-function digit_siteinds(g::AbstractGraph; base=2)
+function digit_siteinds(
```
Sounds good, I can make a file and add some default tag functions.

JoeyT1994 commented:
@ryanlevy I added support for complex delta functions by changing the `delta_p` constructor to not assume that there is only one index per vertex. With that generalization everything seems to work and I added tests

JoeyT1994 commented:
@ryanlevy any further comments or ready to merge?

ryanlevy reviewed (approved):
Thanks!
